rasterio/merge.py

```diff
@@ -392,3 +397,317 @@ def nullcontext(obj):
+        dst_s = math.floor(dst_s / res[1]) * res[1]
+        dst_n = math.ceil(dst_n / res[1]) * res[1]
+
+    # Compute output shape. We guarantee it will cover the output bounds
+    # completely.
+    output_width = int(round((dst_e - dst_w) / res[0]))
+    output_height = int(round((dst_n - dst_s) / res[1]))
+
+    output_transform = Affine.translation(dst_w, dst_n) * Affine.scale(res[0], -res[1])
+
+    if dtype is not None:
+        dt = dtype
+        logger.debug("Set dtype: %s", dt)
+
+    # create destination array
+    # dest = np.zeros((output_count, output_height, output_width), dtype=dt)
+
+    # Create destination VRT.
+    vrtdataset = ET.Element(
+        "VRTDataset",
+        rasterYSize=str(output_height),
+        rasterXSize=str(output_width),
+    )
+
+    ET.SubElement(vrtdataset, "SRS").text = crs_wkt if crs_wkt else ""
+    ET.SubElement(vrtdataset, "GeoTransform").text = ",".join(
+        [str(v) for v in output_transform.to_gdal()]
+    )
+
+    if nodata is not None:
+        # Only fill if the nodataval is within dtype's range
+        inrange = False
+        if np.issubdtype(dt, np.integer):
+            info = np.iinfo(dt)
+            inrange = info.min <= nodata <= info.max
+        elif np.issubdtype(dt, np.floating):
+            if math.isnan(nodata):
+                inrange = True
+            else:
+                info = np.finfo(dt)
+                inrange = info.min <= nodata <= info.max
+        if inrange:
+            nodataval = nodata
+        else:
+            warnings.warn(
+                "The nodata value, %s, is beyond the valid "
+                "range of the chosen data type, %s. Consider overriding it "
+                "using the --nodata option for better results." % (nodataval, dt)
+            )
+    else:
+        nodataval = None
+
+    # Create VRT bands.
+    for bidx, ci, block_shape, dtype in zip(
+        indexes, colorinterps, block_shapes, dtypes
+    ):
+        vrtrasterband = ET.SubElement(
+            vrtdataset,
+            "VRTRasterBand",
+            dataType=_gdal_typename(dtype),
+            band=str(bidx),
+        )
+
+        if background is not None or nodataval is not None:
+            ET.SubElement(vrtrasterband, "NoDataValue").text = str(
+                background or nodataval
+            )
+
+            if hidenodata:
+                ET.SubElement(vrtrasterband, "HideNoDataValue").text = "1"
+
+        ET.SubElement(vrtrasterband, "ColorInterp").text = ci.name.capitalize()
+
+    # Add sources to VRT bands.
+    for idx, dataset in enumerate(datasets):
+        with dataset_opener(dataset) as src:
+            for bidx, ci, block_shape, dtype in zip(
+                src.indexes, src.colorinterp, src.block_shapes, src.dtypes
+            ):
+                vrtrasterband = vrtdataset.find(f"VRTRasterBand[@band='{bidx}']")
+                complexsource = ET.SubElement(vrtrasterband, "ComplexSource")
```

Comment:
if we add a `resampling=resampling.name` keyword argument here, we can support resampling when SrcRect and DstRect differ.

> A resampling attribute can be specified on a SimpleSource or ComplexSource element to specified the resampling algorithm used when the size of the destination rectangle is not the same as the size of the source rectangle. The values allowed for that attribute are : nearest,bilinear,cubic, cubicspline,lanczos,average,mode.

Reply:
That assumes, of course we are passed an enum member. Otherwise we would have to do `Resampling(<int>).name` to get the proper resampling string for the attribute.

Reply:
Reampling is added in 14192e33.